av.h

```diff
@@ -98,54 +98,72 @@ If all you need is to look up an array element, then prefer C<av_fetch>.
 
 /*
 =for apidoc newAV
+=for apidoc_item newAV_alloc_x
+=for apidoc_item newAV_alloc_xz
 
-Creates a new AV.  The reference count is set to 1.
+These all create a new AV, setting the reference count to 1. They differ
+in when and how an accompanying SV* array is allocated and initialized.
 
-Perl equivalent: C<my @array;>.
+The Perl equivalent is approximately C<my @array;>.
 
-=cut
-*/
+newAV does not allocate a SV* array.
+    AV *av = newAV();
 
-#define newAV()	MUTABLE_AV(newSV_type(SVt_PVAV))
+This is very useful when an AV is required, but population of it may be
+deferred, or even never actually take place. (Memory is not allocated
+unnecessarily.)
 
-/*
-=for apidoc newAV_alloc_x
-=for apidoc newAV_alloc_xz
+Subsequent SV* array allocation would be performed via av_extend. This
+might be called directly:
+    av_extend(av, key);
 
-Similar to newAV(), but a SV* array is also allocated.
+Or it might be called implicitly when the first element is stored:
+    (void)av_store(av, 0, sv);
 
-This is similar to but more efficient than doing:
+Unused array elements are typically initialized by av_extend, but this
+is undesirable and will not be the case for some specific arrays.
```

Comment:
Where do they go to find the rules about which specific arrays and what

Reply:
Those specific arrays will be the non-`AvREAL()` ones, which don't seem to be described anywhere. I'm happy to have a go at documenting "real/unreal" arrays (in _perlguts_ or in code comments?)  but would rather not do as part of this PR.

For now, should it just say:
`Unused array elements are typically initialized by av_extend.`?

Reply:
Or we hold this PR until AvREAL is documented, and then go back to the original text of:
```
Whether or not any as-yet-untouched array elements are initialized by
av_extend depends upon whether the array is "real" at the time.
```

Reply:
I take that back, there are some comments at the top of `av.h`. I'll have a look at converting them to apidoc and expanding them.

Reply:
I've converted this sentence to:
```
Unused array elements are typically initialized by C<av_extend>. (Only
core maintainers should have need to concern themseleves with when that
is not the case. Refer to F<av.h> and F<av.c> for the differences between 
real and fake AVs.)
```
And opened https://github.com/Perl/perl5/pull/19062 as a separate concern for the documentation of real/fake AVs.